File: py/src/imbrace/resources/teams.py
from typing import Any, Dict, List, Optional
from ..http import HttpTransport, AsyncHttpTransport


class TeamsResource:
    """Teams domain — Sync.

    @param base - platform service base URL (gateway/platform)
    """

    def __init__(self, http: HttpTransport, base: str):
        self._http = http
        self._base = base.rstrip("/")

    @property
    def _v1(self) -> str:
        return f"{self._base}/v1"

    @property
    def _v2(self) -> str:
        return f"{self._base}/v2"

    def list(self, limit: Optional[int] = None, skip: Optional[int] = None, q: Optional[str] = None) -> Dict[str, Any]:
        params: Dict[str, Any] = {}
        if limit is not None:
            params["limit"] = limit
        if skip is not None:
            params["skip"] = skip
        if q:
            params["q"] = q
        return self._http.request("GET", f"{self._v1}/teams", params=params).json()

    def list_my(self) -> Dict[str, Any]:
        return self._http.request("GET", f"{self._v2}/teams/my").json()

    def create(self, body: Dict[str, Any]) -> Dict[str, Any]:
        return self._http.request("POST", f"{self._v1}/teams", json=body).json()

    def update(self, team_id: str, body: Dict[str, Any]) -> Dict[str, Any]:
        return self._http.request("PUT", f"{self._v2}/teams/{team_id}", json=body).json()

    def delete(self, team_id: str) -> Dict[str, Any]:
        return self._http.request("DELETE", f"{self._v2}/teams/{team_id}").json()

    def add_users(self, team_id: str, user_ids: List[str]) -> Dict[str, Any]:
        return self._http.request("POST", f"{self._v2}/teams/_add_users",
                                  json={"team_id": team_id, "user_ids": user_ids}).json()

    def remove_users(self, user_ids: List[str]) -> Dict[str, Any]:
        return self._http.request("POST", f"{self._v2}/teams/_remove_users",
                                  json={"user_ids": user_ids}).json()

    def get_users(self, team_id: str) -> Dict[str, Any]:
        return self._http.request("GET", f"{self._v1}/team_users",
                                  params={"team_id": team_id}).json()


class AsyncTeamsResource:
    """Teams domain — Async."""

    def __init__(self, http: AsyncHttpTransport, base: str):
        self._http = http
        self._base = base.rstrip("/")

    @property
    def _v1(self) -> str:
        return f"{self._base}/v1"

    @property
    def _v2(self) -> str:
        return f"{self._base}/v2"

    async def list(self, limit: Optional[int] = None, skip: Optional[int] = None, q: Optional[str] = None) -> Dict[str, Any]:
        params: Dict[str, Any] = {}
        if limit is not None:
            params["limit"] = limit
        if skip is not None:
            params["skip"] = skip
        if q:
            params["q"] = q
        res = await self._http.request("GET", f"{self._v1}/teams", params=params)
        return res.json()

    async def list_my(self) -> Dict[str, Any]:
        res = await self._http.request("GET", f"{self._v2}/teams/my")
        return res.json()

    async def create(self, body: Dict[str, Any]) -> Dict[str, Any]:
        res = await self._http.request("POST", f"{self._v1}/teams", json=body)
        return res.json()

    async def update(self, team_id: str, body: Dict[str, Any]) -> Dict[str, Any]:
        res = await self._http.request("PUT", f"{self._v2}/teams/{team_id}", json=body)
        return res.json()

    async def delete(self, team_id: str) -> Dict[str, Any]:
        res = await self._http.request("DELETE", f"{self._v2}/teams/{team_id}")
        return res.json()

    async def add_users(self, team_id: str, user_ids: List[str]) -> Dict[str, Any]:
        res = await self._http.request("POST", f"{self._v2}/teams/_add_users",
                                       json={"team_id": team_id, "user_ids": user_ids})
        return res.json()

File: ts/src/resources/teams.ts
import { HttpTransport } from "../http.js"

export class TeamsResource {
  /**
   * @param base - platform base URL (gateway/platform)
   */
  constructor(private readonly http: HttpTransport, private readonly base: string) {}

  private get v1() { return `${this.base}/v1` }
  private get v2() { return `${this.base}/v2` }

  async list(params?: { type?: string; limit?: number; skip?: number; q?: string }) {
    const url = new URL(`${this.v2}/teams`)
    if (params?.type)   url.searchParams.set("type",  params.type)
    if (params?.limit)  url.searchParams.set("limit", String(params.limit))
    if (params?.skip !== undefined) url.searchParams.set("skip", String(params.skip))
    if (params?.q)      url.searchParams.set("q",     params.q)
    return this.http.getFetch()(url, { method: "GET" }).then(r => r.json())
  }

  async listMy() {
    return this.http.getFetch()(`${this.v2}/teams/my`, { method: "GET" }).then(r => r.json())
  }

  async create(body: Record<string, unknown>) {
    return this.http.getFetch()(`${this.v1}/teams`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async update(teamId: string, body: Record<string, unknown>) {
    return this.http.getFetch()(`${this.v2}/teams/${teamId}`, {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async delete(teamId: string): Promise<void> {
    await this.http.getFetch()(`${this.v2}/teams/${teamId}`, { method: "DELETE" })
  }

  async addUsers(body: { team_id: string; user_ids: string[] }) {
    return this.http.getFetch()(`${this.v2}/teams/_add_users`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async removeUsers(body: { user_ids: string[] }) {
    return this.http.getFetch()(`${this.v2}/teams/_remove_users`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async getUsers(teamId: string) {
    return this.http.getFetch()(`${this.v1}/team/${teamId}/users`, { method: "GET" }).then(r => r.json())
  }

  async getWorkflows(teamId: string) {
    return this.http.getFetch()(`${this.v1}/teams/${teamId}/workflows`, { method: "GET" }).then(r => r.json())
  }

  async join(body: Record<string, unknown>) {
    return this.http.getFetch()(`${this.v2}/teams/_join_team`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async leave(body: Record<string, unknown>) {
    return this.http.getFetch()(`${this.v2}/teams/_leave`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async requestJoin(teamId: string, body: Record<string, unknown>) {
    return this.http.getFetch()(`${this.v2}/teams/${teamId}/join_request`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async approveJoinRequest(teamId: string, teamUserId: string) {
    return this.http.getFetch()(`${this.v2}/teams/${teamId}/user/${teamUserId}/approve`, {
      method: "POST",
    }).then(r => r.json())
  }

  async updateUserRole(teamId: string, teamUserId: string, body: Record<string, unknown>) {
    return this.http.getFetch()(`${this.v2}/teams/${teamId}/user/${teamUserId}/role`, {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }
}
